chainbench/tools/discovery/rpc.py

```diff
@@ -0,0 +1,69 @@
+import json
+
+import httpx
+
+
+class JSONRPCError(Exception):
+    pass
+
+
+class RPCDiscovery:
+    @staticmethod
+    def _get_methods_list(clients: list[str]) -> list[str]:
+        methods = []
+        with open("chainbench/tools/discovery/methods.json", "r") as f:
+            methods_json = json.loads(f.read())
+            for method in methods_json.keys():
+                for client in clients:
+                    if client in methods_json[method]["clients"]:
+                        methods.append(method)
+        return methods
+
+    @staticmethod
+    def get_clients_and_versions() -> list[tuple[str, str]]:
+        clients = []
+        with open("chainbench/tools/discovery/clients.json", "r") as f:
+            clients_json = json.loads(f.read())
+            for client in clients_json.keys():
+                if "endpoints" not in clients_json[client]:
+                    clients.append((client, clients_json[client]["version"]))
+                else:
+                    for endpoint in clients_json[client]["endpoints"]:
+                        clients.append((f"{client}{endpoint}", clients_json[client]["version"]))
+        return clients
+
+    @classmethod
+    def discover_methods(cls, endpoint: str, clients: list[str]) -> list[tuple[str, str | bool]]:
```

Comment:
lets have a dataclass with fields method, result, message rather than tuple

Reply:
1) actually I would insist on having dataclasses for all return types here instead of tuples

2) also having separate functions that parse the files first into some types before processing could be easier to understand and modify in future